### ISSUE TITLE
Segmentation Fault with 4.0.0

### ISSUE THREAD

**Issue by normelton** [sinatra/sinatra#2004]
Gems installed:

sinatra 4.0.0
webrick 1.8.1
rack 3.0.10
rack-session 2.0.0
rack-protection 4.0.0
rackup 2.1.0

When I POST the user to /auth/developer, I get the following segmentation fault.
Downgrading to Sinatra 3.2 and the segfault goes away and the app behaves properly.

```
/usr/local/bundle/gems/rackup-2.1.0/lib/rackup/handler/webrick.rb:87: [BUG] Segmentation fault at 0x0033ffff98f706c0
ruby 3.3.0 (2023-12-25 revision 5124f9ac75) [aarch64-linux]

-- Control frame information -----------------------------------------------
c:0037 p:---- s:0258 e:000257 CFUNC  :resume
c:0036 p:0008 s:0254 e:000253 METHOD /usr/local/bundle/gems/rackup-2.1.0/lib/rackup/handler/webrick.rb:87
c:0035 p:0150 s:0250 e:000249 METHOD /usr/local/bundle/gems/rackup-2.1.0/lib/rackup/stream.rb:65
c:0034 p:0116 s:0243 e:000242 METHOD /usr/local/bundle/gems/rack-3.0.10/lib/rack/lint.rb:456
c:0033 p:0106 s:0237 e:000236 METHOD /usr/local/bundle/gems/rack-3.0.10/lib/rack/request.rb:520
c:0032 p:0006 s:0229 e:000227 METHOD /usr/local/bundle/gems/rack-3.0.10/lib/rack/request.rb:546
c:0031 p:0011 s:0224 e:000223 METHOD /usr/local/bundle/gems/rack-3.0.10/lib/rack/request.rb:68
c:0030 p:0048 s:0220 e:000217 METHOD /usr/local/bundle/gems/rack-protection-4.0.0/lib/rack/protection/authenticity_token.rb:119
c:0029 p:0005 s:0212 e:000211 METHOD /usr/local/bundle/gems/rack-protection-4.0.0/lib/rack/protection/base.rb:49
c:0028 p:0026 s:0206 e:000205 METHOD /usr/local/bundle/gems/rack-session-2.0.0/lib/rack/session/abstract/id.rb:272
c:0027 p:0005 s:0195 e:000194 METHOD /usr/local/bundle/gems/rack-session-2.0.0/lib/rack/session/abstract/id.rb:266
c:0026 p:0007 s:0190 e:000189 METHOD /usr/local/bundle/gems/rack-protection-4.0.0/lib/rack/protection/xss_header.rb:20
c:0025 p:0039 s:0182 e:000181 METHOD /usr/local/bundle/gems/rack-protection-4.0.0/lib/rack/protection/path_traversal.rb:18
c:0024 p:0015 s:0176 e:000175 METHOD /usr/local/bundle/gems/rack-protection-4.0.0/lib/rack/protection/json_csrf.rb:28
c:0023 p:0033 s:0167 e:000166 METHOD /usr/local/bundle/gems/rack-protection-4.0.0/lib/rack/protection/base.rb:53
c:0022 p:0033 s:0161 e:000160 METHOD /usr/local/bundle/gems/rack-protection-4.0.0/lib/rack/protection/base.rb:53
c:0021 p:0007 s:0155 e:000154 METHOD /usr/local/bundle/gems/rack-protection-4.0.0/lib/rack/protection/frame_options.rb:33
c:0020 p:0015 s:0147 e:000146 METHOD /usr/local/bundle/gems/rack-3.0.10/lib/rack/null_logger.rb:13
c:0019 p:0007 s:0142 e:000141 METHOD /usr/local/bundle/gems/rack-3.0.10/lib/rack/head.rb:15
c:0018 p:0007 s:0134 e:000133 METHOD /usr/local/bundle/gems/sinatra-4.0.0/lib/sinatra/show_exceptions.rb:23
c:0017 p:0007 s:0125 e:000124 METHOD /usr/local/bundle/gems/sinatra-4.0.0/lib/sinatra/base.rb:224
c:0016 p:0007 s:0118 e:000117 METHOD /usr/local/bundle/gems/sinatra-4.0.0/lib/sinatra/base.rb:2115
c:0015 p:0007 s:0113 e:000112 BLOCK  /usr/local/bundle/gems/sinatra-4.0.0/lib/sinatra/base.rb:1674
c:0014 p:0017 s:0110 e:000109 METHOD /usr/local/bundle/gems/sinatra-4.0.0/lib/sinatra/base.rb:1890
c:0013 p:0004 s:0105 e:000104 METHOD /usr/local/bundle/gems/sinatra-4.0.0/lib/sinatra/base.rb:1674
c:0012 p:0028 s:0100 e:000099 METHOD /usr/local/bundle/gems/rack-3.0.10/lib/rack/tempfile_reaper.rb:20
c:0011 p:0066 s:0092 e:000091 METHOD /usr/local/bundle/gems/rack-3.0.10/lib/rack/lint.rb:63
c:0010 p:0014 s:0087 e:000086 METHOD /usr/local/bundle/gems/rack-3.0.10/lib/rack/lint.rb:35
c:0009 p:0007 s:0082 e:000081 METHOD /usr/local/bundle/gems/rack-3.0.10/lib/rack/show_exceptions.rb:27
c:0008 p:0013 s:0073 e:000072 METHOD /usr/local/bundle/gems/rack-3.0.10/lib/rack/common_logger.rb:43
c:0007 p:0014 s:0063 e:000062 METHOD /usr/local/bundle/gems/sinatra-4.0.0/lib/sinatra/base.rb:266
c:0006 p:0007 s:0058 e:000057 METHOD /usr/local/bundle/gems/rack-3.0.10/lib/rack/content_length.rb:20
c:0005 p:0173 s:0049 e:000048 METHOD /usr/local/bundle/gems/rackup-2.1.0/lib/rackup/handler/webrick.rb:111
c:0004 p:0146 s:0032 e:000031 METHOD /usr/local/bundle/gems/webrick-1.8.1/lib/webrick/httpserver.rb:140
c:0003 p:0235 s:0021 e:000020 METHOD /usr/local/bundle/gems/webrick-1.8.1/lib/webrick/httpserver.rb:96
c:0002 p:0107 s:0009 e:000008 BLOCK  /usr/local/bundle/gems/webrick-1.8.1/lib/webrick/server.rb:310 [FINISH]
c:0001 p:---- s:0003 e:000002 DUMMY  [FINISH]

-- Ruby level backtrace information ----------------------------------------
/usr/local/bundle/gems/webrick-1.8.1/lib/webrick/server.rb:310:in `block in start_thread'
/usr/local/bundle/gems/webrick-1.8.1/lib/webrick/httpserver.rb:96:in `run'
/usr/local/bundle/gems/webrick-1.8.1/lib/webrick/httpserver.rb:140:in `service'
/usr/local/bundle/gems/rackup-2.1.0/lib/rackup/handler/webrick.rb:111:in `service'
/usr/local/bundle/gems/rack-3.0.10/lib/rack/content_length.rb:20:in `call'
/usr/local/bundle/gems/sinatra-4.0.0/lib/sinatra/base.rb:266:in `call'
/usr/local/bundle/gems/rack-3.0.10/lib/rack/common_logger.rb:43:in `call'
/usr/local/bundle/gems/rack-3.0.10/lib/rack/show_exceptions.rb:27:in `call'
/usr/local/bundle/gems/rack-3.0.10/lib/rack/lint.rb:35:in `call'
/usr/local/bundle/gems/rack-3.0.10/lib/rack/lint.rb:63:in `response'
/usr/local/bundle/gems/rack-3.0.10/lib/rack/tempfile_reaper.rb:20:in `call'
/usr/local/bundle/gems/sinatra-4.0.0/lib/sinatra/base.rb:1674:in `call'
/usr/local/bundle/gems/sinatra-4.0.0/lib/sinatra/base.rb:1890:in `synchronize'
/usr/local/bundle/gems/sinatra-4.0.0/lib/sinatra/base.rb:1674:in `block in call'
/usr/local/bundle/gems/sinatra-4.0.0/lib/sinatra/base.rb:2115:in `call'
/usr/local/bundle/gems/sinatra-4.0.0/lib/sinatra/base.rb:224:in `call'
/usr/local/bundle/gems/sinatra-4.0.0/lib/sinatra/show_exceptions.rb:23:in `call'
/usr/local/bundle/gems/rack-3.0.10/lib/rack/head.rb:15:in `call'
/usr/local/bundle/gems/rack-3.0.10/lib/rack/null_logger.rb:13:in `call'
/usr/local/bundle/gems/rack-protection-4.0.0/lib/rack/protection/frame_options.rb:33:in `call'
/usr/local/bundle/gems/rack-protection-4.0.0/lib/rack/protection/base.rb:53:in `call'
/usr/local/bundle/gems/rack-protection-4.0.0/lib/rack/protection/base.rb:53:in `call'
/usr/local/bundle/gems/rack-protection-4.0.0/lib/rack/protection/json_csrf.rb:28:in `call'
/usr/local/bundle/gems/rack-protection-4.0.0/lib/rack/protection/path_traversal.rb:18:in `call'
/usr/local/bundle/gems/rack-protection-4.0.0/lib/rack/protection/xss_header.rb:20:in `call'
/usr/local/bundle/gems/rack-session-2.0.0/lib/rack/session/abstract/id.rb:266:in `call'
/usr/local/bundle/gems/rack-session-2.0.0/lib/rack/session/abstract/id.rb:272:in `context'
/usr/local/bundle/gems/rack-protection-4.0.0/lib/rack/protection/base.rb:49:in `call'
/usr/local/bundle/gems/rack-protection-4.0.0/lib/rack/protection/authenticity_token.rb:119:in `accepts?'
/usr/local/bundle/gems/rack-3.0.10/lib/rack/request.rb:68:in `params'
/usr/local/bundle/gems/rack-3.0.10/lib/rack/request.rb:546:in `params'
/usr/local/bundle/gems/rack-3.0.10/lib/rack/request.rb:520:in `POST'
/usr/local/bundle/gems/rack-3.0.10/lib/rack/lint.rb:456:in `read'
/usr/local/bundle/gems/rackup-2.1.0/lib/rackup/stream.rb:65:in `read'
/usr/local/bundle/gems/rackup-2.1.0/lib/rackup/handler/webrick.rb:87:in `read_next'
/usr/local/bundle/gems/rackup-2.1.0/lib/rackup/handler/webrick.rb:87:in `resume'

-- Threading information ---------------------------------------------------
Total ractor count: 1
Ruby thread count for this ractor: 3

-- Machine register context ------------------------------------------------
  x0: 0x0000aaaac7951ee0  x1: 0x0000ffff7811eac0  x2: 0x0000ffff7c1ae160
  x3: 0x0000ffff7ec6ff60  x4: 0x0000ffff7ec70018  x5: 0x0000ffff7ec90000
  x6: 0x0000ffff994c7b88  x7: 0x0000000000000000 x18: 0x0000aaaac7066270
 x19: 0x0000000000000000 x20: 0x0000000000000000 x21: 0x0000000000000000
 x22: 0x0000000000000000 x23: 0x0000000000000000 x24: 0x0000000000000000
 x25: 0x0000000000000000 x26: 0x0000000000000000 x27: 0x0000000000000000
 x28: 0x0000000000000000 x29: 0x0000000000000000  sp: 0x0000ffff7ec70000
 fau: 0x0033ffff98f706c0

-- C level backtrace information -------------------------------------------
Segmentation fault
```

**Comment by dentarg** [comment]
What if you try the same with Ruby 3.2?

**Comment by normelton** [comment]
Interesting. This is all in docker containers. If I use ruby:latest, it crashes. With ruby:3.2, it works fine.

But it also works fine with ruby:3.3. I don't understand the difference between ruby:3.3 and ruby:latest, but will do some more digging on this tonight.

What led you to suspect Ruby 3.3 / 3.2?

**Comment by dentarg** [comment]
Segfaults are always a Ruby bug. You should report it to the Ruby project. They will appreciate a minimal reproduction (preferably without Sinatra).

**Comment by dentarg** [comment]
(You may have ruby:latest cached, please double check)

**Comment by normelton** [comment]
> (You may have ruby:latest cached, please double check)

Oh. Duh. The problem is already fixed in the newest version of 3.3. You were correct, I had an older version cached. Thank you!
